dialects: FSM advanced operator (SW/HW instances)

luisacicolini opened the pull request:
this PR is built on top of #1600 and integrates the base operators and testing with advanced ones useful to emulate FSM behavior in hw or sw.

compor commented:
Can you please stack it on top of #1600? it'd be easier to review

compor commented on xdsl/dialects/fsm.py:
```diff
@@ -45,6 +47,11 @@
 
+@irdl_attr_definition
+class InstanceType(ParametrizedAttribute, TypeAttribute):
+    name = "fsm.instancetype"
```
Could you add some doc of this type, please?

luisacicolini replied:
Unfortunately I was not able to find any. I managed to define it by basically looking at .cpp and .td files, as the documentation refers to it was "FSM instance type" :(

compor commented on xdsl/dialects/fsm.py:
```diff
@@ -30,9 +30,11 @@
     var_operand_def,
     var_result_def,
 )
+from xdsl.irdl.irdl import irdl_attr_definition
```
```suggestion
from xdsl.irdl import irdl_attr_definition
```

compor commented on xdsl/dialects/fsm.py:
```diff
@@ -363,6 +372,156 @@ def __init__(
+@irdl_op_definition
+class Instance(IRDLOperation):
+    """Represents an instance of a state machine, including an
+    instance name and a symbol reference of the machine"""
```
```suggestion
    """
    Represents an instance of a state machine, including an
    instance name and a symbol reference of the machine.
    """
    
    Similar in other docstrings.
```

compor commented on xdsl/dialects/fsm.py:
```diff
@@ -363,6 +372,156 @@ def __init__(
 
 
+@irdl_op_definition
+class Instance(IRDLOperation):
```
```suggestion
class InstanceOp(IRDLOperation):
```

likewise below.

compor commented on xdsl/dialects/fsm.py:
```diff
@@ -200,11 +209,11 @@ def verify_(self):
                 ):
                     raise VerifyException(
-                        "Output types must be consistent with the machine's"
+                        "Output type must be consistent with the machine's"
```
```suggestion
                        "Output type must match with the containing state machine's output types"
```

I know that the doc says "consistent", but that means to be the same/matching, right?
If so, reword messages below too?

luisacicolini replied:
correct, it basically needs to be the same type as the output type of the function defined in the MachineOp

compor commented on xdsl/dialects/fsm.py:
```diff
@@ -363,6 +372,156 @@ def __init__(
+                and len(self.inputs) == len(m.function_type.inputs)
+            ):
+                raise VerifyException(
+                    "Inputs types must be consistent with the machine's "
```
Could you add the state machine's `sym_name` in these messages? Could it help debugging by knowing to which machine this error refers too based on the symbol table?

If so, adjust in the other similar cases where a symbol table lookup is involved too?

JosseVanDelm commented:
Some files look like they were already reviewed in #1600 ?

luisacicolini commented:
Thank you everyone for your suggestions :) 
I added new more significant strings for exceptions and rebased the branch to main (after merging #1600). 
Please let me know if there are any major changes I should implement!

superlopuh commented:
There are still some merge conflicts left over, could you please resolve them first?

compor commented:
@luisacicolini can you ping us when you resolve the conflicts and everything on this patch is properly based on `main` please?

luisacicolini commented:
Manually checked the files again this morning, should be okay!

compor commented on xdsl/dialects/fsm.py:
```diff
@@ -228,6 +232,8 @@ class TransitionOp(IRDLOperation):
 
     action = region_def()
 
+    # attributes
```
```suggestion
```

compor commented on xdsl/dialects/fsm.py:
```diff
@@ -371,6 +372,176 @@ def __init__(
+        if not isinstance(self.instance.owner, InstanceOp):
+            raise VerifyException("The instance operand must be Instance")
+
+        # operand types must match the machine input types
+        # result types must match the machine output types
```
```suggestion
```

compor reviewed (approved):
LGTM with some nits

math-fehr commented on xdsl/dialects/fsm.py:
```diff
@@ -297,20 +299,22 @@ def verify_(self) -> None:
-            )
-        elif found > 1:
-            raise VerifyException(
-                "Multiple updates to the same variable within a single action region is disallowed"
-            )
+        while parent is not None:
+            if isinstance(parent, TransitionOp):
+                # walk through the action region
+                found = 0
+                for op in parent.action.walk():
+                    if isinstance(op, UpdateOp) and op.variable == self.variable:
+                        found += 1
+                if found == 0:
+                    raise VerifyException(
+                        "Update must only be located in the action region of a transition"
+                    )
+                elif found > 1:
+                    raise VerifyException(
+                        "Multiple updates to the same variable within a single action region is disallowed"
+                    )
+            parent = parent.parent_op()
```
I feel that this change is backwards, maybe it's a merge problem?
If you want I could take a look with you tomorrow

luisacicolini replied:
yup I guess you're right, I had quite some issues with main merging-rebasing so I guess I missed that! Sure thing :)

math-fehr commented on xdsl/dialects/fsm.py:
```diff
@@ -195,9 +196,11 @@ def __init__(
 
     def verify_(self):
         parent = self.parent_op()
-        assert isinstance(parent, StateOp)
```
Same here, this feels backwards?

math-fehr reviewed (approved):
Really nice!
I think some of these changes appeared while you were doing your rebase?
Because some of the changes in the earlier operations are actually old code I believe?
